### PR TITLE
Fix/3451  strict user config validation

### PR DESCRIPTION
**Pull request by inkhey**
related to #3451 

## Checkpoints

These points must be checked before merging. Please don't edit them out.

**For developers**

- [X] If relevant, manual tests have been done to ensure the stability of the whole application and that the involved feature works
- [x] The original issue is up to date w.r.t the latest discussions and contains a short summary of the implemented solution
- [X] Automated tests covering the feature or the fix, have been written, deemed irrelevant (give the reason), or an issue has been created to implement the test (give the link)

**For code reviewers**

- [x] The code is clear enough
- [x] If there are FIXMEs in the code, related issues are mentioned in the FIXME
- [x] If there are TODOs, NOTEs or HACKs in code, the date and the developer initials are present

**For testers**

- [x] Manual, quality tests have been done

### CODE REVIEW

File: backend/tracim_backend/views/core_api/schemas.py [hunk 1]
```diff
@@ -266,7 +267,20 @@ class LoggedInUserPasswordSchema(marshmallow.Schema):
...
+
+    parameters = marshmallow.fields.Dict(
+        required=True,
+        example={"param1": "value1"},
+        validate=user_config_validator,
+        description="A simple json dictionary. "
+        'Valid key are limited to not empty string with "0-9a-zA-Z-_." characters. '
+        'You can use "." to create hierarchy in the configuration parameters. '
+        "Valid value allow only standard type: int, bool, null, float and do not accept "
+        "complex type such dict or list.",
```
Comment (raphj) [inline backend/tracim_backend/views/core_api/schemas.py:282]:
```suggestion
```suggestion
        'Valid keys only contain characters in "0-9a-zA-Z-_." and are not empty. '
        'You can use "." to create a hierarchy in the configuration parameters. '
        "Valid values only allow primitive types: numbers, bool, null, and do not accept "
        "complex types such dictionaries or lists.",
```

File: backend/tracim_backend/views/core_api/user_controller.py [hunk 2]
```diff
@@ -746,7 +746,9 @@ def get_user_config(self, context, request: TracimRequest, hapic_data: HapicData
     @hapic.output_body(NoContentSchema(), default_http_code=HTTPStatus.NO_CONTENT)
     def set_user_config(self, context, request: TracimRequest, hapic_data: HapicData) -> None:
         """
-        set or update the given configuration parameters for the given user
+        Set or update the given configuration parameters for the given user
+        The behavior of this endpoint is adding/updating key (patch-like) but not replacing the
+        whole config, so it's not possible to remove key through this endpoint.
```
Comment (raphj) [inline backend/tracim_backend/views/core_api/user_controller.py:751]:
```suggestion
        whole configuration, so it's not possible to remove keys through this endpoint.
```

File: backend/tracim_backend/lib/utils/utils.py [hunk 3]
```diff
@@ -499,11 +502,42 @@ def get_build_version(path: str) -> str:
...
+            raise ValidationError('Dictionary key "{}" is not a string'.format(key))
+        regex_validator = Regexp(regex=(re.compile("^[0-9a-zA-Z-_.]+$")))
+        try:
+            regex_validator(key)
+        except ValidationError as exc:
+            raise ValidationError(
+                'Dictionary key "{}" incorrect : {}'.format(key, str(exc))
+            ) from exc
+
+    # INFO - G.M - We do consider float is the type used for float conversion,
```
Comment (raphj) [inline backend/tracim_backend/lib/utils/utils.py:532]:
```suggestion
    # INFO - G.M - We asume float is the type used for float conversion,
```

**Review by raphj — CHANGES_REQUESTED** [review]
spelling, otherwise lgtm

File: backend/tracim_backend/lib/utils/utils.py (hunk 3 above)
Comment (grignards) [reply backend/tracim_backend/lib/utils/utils.py:532]: assume…

File: backend/tracim_backend/lib/utils/utils.py [hunk 4]
```diff
@@ -499,11 +502,42 @@ def get_build_version(path: str) -> str:
...
+    # this may change depending
+    # on how the json parser is configured.
+    float_type = float
+    valid_types = [str, int, bool, float_type]
+    for value in dict_.values():
+        if value and type(value) not in valid_types:
```
Comment (grignards) [inline backend/tracim_backend/lib/utils/utils.py:538]: Could be easier by using `isinstance`:
```python
for value in dict_.values():
    if not isinstance(value, (str, int, bool, float, type(None))):
        ...
```
And more "user-friendly" by collecting all the invalid keys before raising the exception, eg.
```python
invalid_key_value_pairs = [
    (key, value)
    for key, value in dict_.items()
    if not isinstance(value, (str, int, bool, float, type(None)))
]
if invalid_key_value_pairs:
   raise ValidationError('Only string/number/null values are allowed. Invalid values: {}'.format(invalid_key_value_pairs))
```

Comment (inkhey) [reply backend/tracim_backend/lib/utils/utils.py:538]: Oh nice, i didin't know that isinstance could take a list of type as parameters.

**Review by grignards — APPROVED** [review]

**Review by raphj — APPROVED** [review]

**Review by grignards — APPROVED** [review]

**Comment by inkhey** [comment]
@gbonaspetti 
can you verify if GET /api/users/1/config work for admin@admin;admin default user from a fresh database using this branch ?

I don't think other manual tests are required here.

**Comment by gbonaspetti** [comment]
> @gbonaspetti
> can you verify if GET /api/users/1/config work for admin@admin;admin default user from a fresh database using this branch ?
> 
> I don't think other manual tests are required here.

Tested this on branch.

**Review by raphj — APPROVED** [review]

**Pull request merged**